File: src/services/paystack.ts


import axios, { AxiosError } from 'axios';
import { config } from '../config';
import { PaystackResolveResponse, PaystackBankListResponse } from '../types';
import { NameEnquiryFailedError, NameEnquiryNoMatchError } from '../errors';

const PAYSTACK_BASE = 'https://api.paystack.co';

const paystackClient = axios.create({
  baseURL: PAYSTACK_BASE,
  timeout: 10_000,
  headers: {
    Authorization: `Bearer ${config.paystackSecretKey}`,
  },
});

export async function resolveAccount(
  accountNumber: string,
  bankCode: string
): Promise<{ name: string; accountNumber: string }> {
  try {
    const response = await paystackClient.get<PaystackResolveResponse>('/bank/resolve', {
      params: {
        account_number: accountNumber,
        bank_code: bankCode,
      },
    });

    if (
      response.data.status &&
      response.data.data &&
      response.data.data.account_number === accountNumber &&
      response.data.data.account_name.trim().length > 0
    ) {
      return {
        name: response.data.data.account_name.trim(),
        accountNumber: response.data.data.account_number,
      };
    }

    throw new NameEnquiryNoMatchError(accountNumber, bankCode);
  } catch (error) {
    if (error instanceof NameEnquiryNoMatchError) {
      throw error;
    }

    const axiosError = error as AxiosError<{ message?: string }>;

    if (axiosError.response?.status === 422) {
      throw new NameEnquiryNoMatchError(accountNumber, bankCode);
    }

    throw new NameEnquiryFailedError({
      account_number: accountNumber,
      bank_code: bankCode,
      paystack_error: axiosError.response?.data?.message || axiosError.message,
    });
  }
}

export async function fetchBankList(): Promise<PaystackBankListResponse['data']> {
  try {
    const response = await paystackClient.get<PaystackBankListResponse>('/bank', {
      params: {
        country: 'nigeria',
        perPage: 100,
      },
    });

    if (response.data.status && response.data.data) {
      return response.data.data;
    }

    return [];
  } catch (error) {
    console.error('[Paystack] Failed to fetch bank list:', (error as Error).message);
    return [];
  }
}

File: snapaccount-py/services/paystack.py


import httpx
from config import PAYSTACK_SECRET_KEY

PAYSTACK_BASE = "https://api.paystack.co"

def _headers() -> dict[str, str]:
    return {"Authorization": f"Bearer {PAYSTACK_SECRET_KEY}"}

async def resolve_account(
    account_number: str, bank_code: str
) -> dict:
    """
    Resolve an account number + bank code into the account holder's name.
    Uses Paystack's free /bank/resolve endpoint.

    Returns: {"account_name": "...", "account_number": "..."}
    Raises: ValueError on invalid account, RuntimeError on failure.
    """
    async with httpx.AsyncClient(timeout=10.0) as client:
        try:
            resp = await client.get(
                f"{PAYSTACK_BASE}/bank/resolve",
                params={"account_number": account_number, "bank_code": bank_code},
                headers=_headers(),
            )

            if resp.status_code == 200:
                data = resp.json()
                if data.get("status") and data.get("data"):
                    return {
                        "account_name": data["data"]["account_name"],
                        "account_number": data["data"]["account_number"],
                    }

            if resp.status_code == 422:
                raise ValueError(
                    f"Account not found: {account_number} at bank {bank_code}"
                )

            raise RuntimeError(
                f"Paystack error {resp.status_code}: {resp.text[:200]}"
            )

        except httpx.TimeoutException:
            raise RuntimeError("Paystack request timed out")
        except httpx.HTTPError as e:
            raise RuntimeError(f"Paystack HTTP error: {e}")

async def fetch_bank_list() -> list[dict]:
    """
    Fetch the list of Nigerian banks from Paystack.
    Returns raw bank data — the bank_registry module caches this.
    """
    async with httpx.AsyncClient(timeout=10.0) as client:
        try:
            resp = await client.get(
                f"{PAYSTACK_BASE}/bank",
                params={"country": "nigeria", "perPage": 100},
                headers=_headers(),
            )

            if resp.status_code == 200:
                data = resp.json()
                if data.get("status") and data.get("data"):
                    return data["data"]

            return []

        except Exception as e:
            print(f"[Paystack] Failed to fetch bank list: {e}")
            return []
